File: chapters/14/python/aggregation.py
import pymongo

client = pymongo.MongoClient("localhost", 27017)

# set our database to "cooker"
db = client.cooker
collection = db["recipes"]

pipeline = [
    {
        "$group": {
            "_id": "$type",
            "count": {"$sum": 1}
        }
    },
    {
        "$sort": {"count": -1}
    }
]

result = list(collection.aggregate(pipeline))

for doc in result:
    print(doc["_id"], "has", doc["count"], "recipe(s)")

File: chapters/14/node/aggregation.js
const MongoClient = require('mongodb').MongoClient;

const uri = 'mongodb://localhost:27017';
const db = 'website';

(async function () {
  const client = new MongoClient(uri, { useUnifiedTopology: true });

  try {
    await client.connect();
    const database = client.db(db);
    const collection = database.collection('cookbook');

    const pipeline = [
      { '$group': { '_id': "$type", 'count': { '$sum': 1 } } },
      { '$sort': { 'count': -1 } }
    ];

    const cursor = collection.aggregate(pipeline);

    // iterate over the cursor
    while (await cursor.hasNext()) {
      const doc = await cursor.next();
      console.log(`${doc._id} has ${doc.count} recipe(s)`);
    }

  } catch (err) {
    console.log(err.stack);
  }

  await client.close();
})();
